onmt/decoders/decoder.py

```diff
@@ -104,7 +104,7 @@ def __init__(self, rnn_type, bidirectional_encoder, num_layers,
         self._reuse_copy_attn = reuse_copy_attn
 
     def forward(self, tgt, memory_bank, state, memory_lengths=None,
-                step=None):
+                step=None, fast=None):
```

Comment:
Maybe @srush have a comment on adding this argument. For the context, it was apparently too costly to update the cache in `beam_update` so this feature expects translating with `-fast`.

If we go this way, all decoder classes should be updated. Related to https://github.com/OpenNMT/OpenNMT-py/pull/793.

Reply:
Hmm, yeah, I would really like these classes not to know about decoding, particularly in forward. If you have this option, maybe give it a more general now. Alternatively, you could have it be a flag on the module itself.

Reply:
Thank you both for your help. I added a `with_cache` argument to the `init_decoder_state` function, which is put to `True` only when going through `_fast_translate_batch`. Hope this fits better your expectations @srush !